File: lintcode/reverse-linked-list-ii.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/reverse-linked-list-ii/

from linked_list import *

"""
Definition of ListNode

class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The head of linked list
    @param m: start position
    @param n: end position
    """
    def reverseBetween(self, head, m, n):
        # write your code here
        dummy = ListNode(0)
        writer = dummy
        cursor = head
        prev_m = dummy

        i = 1
        while cursor:
            if i <= m or i > n:
                prev_m = writer
                next = cursor.next
                cursor.next = None
                writer.next = cursor
                writer = cursor
                cursor = next
            elif i > m and i <= n:
                next = cursor.next
                cursor.next = prev_m.next
                prev_m.next = cursor
                cursor = next
            i += 1

        return dummy.next


if __name__ == '__main__':
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)

    print_linked_list(head)
    s = Solution()
    new_head = s.reverseBetween(head, 1, 5)
    print_linked_list(new_head)

File: leetcode/reverse-linked-list-ii.cc
// Question: https://leetcode.com/problems/reverse-linked-list-ii/

#include "headers.h"

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        ListNode dummy(0);
        ListNode* c1 = dummy.next;
        ListNode* c2 = &dummy;
        int i = 0;
        while (head) {
            ++i;
            ListNode* next = head->next;
            if (i <= m || i > n) {
                c2->next = head;
                c1 = c2;
                c2 = c2->next;
                c2->next = NULL;
            } else if (i > m && i <= n) {
                head->next = c1->next;
                c1->next = head;
            }
            head = next;
        }
        return dummy.next;
    }
};

int main(int argc, char** argv) {
    Solution s;
    ListNode* head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);
    head->next->next->next->next = new ListNode(5);
    ListNode* new_head = s.reverseBetween(head, 2, 4);
    while (new_head) {
        cout << new_head->val << " ";
        new_head = new_head->next;
    }
    cout << endl;
    return 0;
}
